pkg/autoscaler/stats_scraper.go

```diff
@@ -52,48 +60,119 @@ var cacheDisabledClient = &http.Client{
 	Timeout: httpClientTimeout,
 }
 
+type readyPods struct {
+	time  time.Time
+	count int
+}
+
 // ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
 // pod to be picked up to serve the request is decided by K8S. Please see
 // https://kubernetes.io/docs/concepts/services-networking/network-policies/
 // for details.
 type ServiceScraper struct {
-	httpClient *http.Client
-	url        string
-	metricKey  string
-	logger     *zap.SugaredLogger
+	httpClient      *http.Client
+	endpointsLister corev1listers.EndpointsLister
+	url             string
+	namespace       string
+	revisionService string
+	metricKey       string
+	cachedReadyPods readyPods
 }
 
 // NewServiceScraper creates a new StatsScraper for the Revision which
 // the given Metric is responsible for.
-func NewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (*ServiceScraper, error) {
-	return newServiceScraperWithClient(metric, logger, cacheDisabledClient)
+func NewServiceScraper(metric *Metric, dynamicConfig *DynamicConfig, endpointsInformer corev1informers.EndpointsInformer) (*ServiceScraper, error) {
+	return newServiceScraperWithClient(metric, dynamicConfig, endpointsInformer, cacheDisabledClient)
 }
 
-func newServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (*ServiceScraper, error) {
+func newServiceScraperWithClient(
+	metric *Metric,
+	dynamicConfig *DynamicConfig,
+	endpointsInformer corev1informers.EndpointsInformer,
+	httpClient *http.Client) (*ServiceScraper, error) {
+	if metric == nil {
+		return nil, errors.New("empty point of Metric")
```

Comment:
```suggestion
		return nil, errors.New("metric must not be nil")
```

This applies to all the error messages. These are all empty pointers technically. But "empty point of Metric" would not tell me what went wrong here.